docs/Advanced-Concepts/CDEs.rst

```diff
@@ -0,0 +1,108 @@
+.. _cdes:
+
+Context-Dependent-Environments
+========================================
+
+Readers may notice that the parameterization system frequently uses ``(site, here, up)``.
+This construct is an artifact of the "context-dependent-environment" system which Chipyard and Rocket Chip both leverage for powerful composable hardware configuration.
+
+The CDE parameterization system provides different "Views" of a single global parameterization. The syntax for accessing a ``Field`` within a ``View`` is ``my_view(MyKey, site_view)``, where ``site_view`` is a "global" view that will be passed recursively into various functions and key-lookups in the call-stack of ``my_view(MyKey, site_view)``.
+
+.. note::
+   Rocket Chip based designs will frequently use ``val p: Parameters`` and ``p(SomeKey)`` to lookup the value of a key. ``Parameters`` is just a subclass of the ``View`` abstract class, and ``p(SomeKey)`` really expands into ``p(SomeKey, p)``. This is because we consider the call ``p(SomeKey)`` to be the "site", or "source" of the original key query, so we need to pass in the view of the configuration provided by ``p`` recursively to future calls through the ``site`` argument.
+
+Consider the following example using CDEs.
+
+.. code:: scala
+
+    case object SomeKeyX extends Field[Boolean](false) // default is false
+    case object SomeKeyY extends Field[Boolean](false) // default is false
+    case object SomeKeyZ extends Field[Boolean](false) // default is false
+
+    class WithX(b: Boolean) extends Config((site, here, up) => {
+      case SomeKeyX => b
+    }
+
+    class WithY(b: Boolean) extends Config((site, here, up) => {
+      case SomeKeyY => b
+    }
+
+
+When forming a query based on a ``Parameters`` object, like ``p(SomeKeyX)``, the configuration system traverses the "chain" of mixins until it finds a partial function which is defined at the key, and then returns that value.
+
+.. code:: scala
+
+    val params = Config(new WithX(true) ++ new WithY(true)) // "chain" together mixins
+    params(SomeKeyX) // evaluates to true
+    params(SomeKeyY) // evaluates to true
+    params(SomeKeyZ) // evaluates to false
+
+In this example, the evaluation of ``params(SomeKeyX)`` will terminate in the partial function defined in ``WithX(true)``, while the evaluation of ``params(SomeKeyY)`` will terminate in the partial function defined in ``WithY(true)``. Note that when no partial functions match, the evaluation will return the default value for that parameter.
+
+The real power of CDEs arises from the ``(site, here, up)`` parameters to the partial functions, which provide useful "views" into the global parameterization that the partial functions may access to determine a parameterization.
+
```

Comment:
"Additional information about the reasons behind CDE can be found in chapter 2 of https://www2.eecs.berkeley.edu/Pubs/TechRpts/2016/EECS-2016-89.pdf"